File: Java/merge_sort.java
// Java program for Merge Sort

class merge_sort {
  
    // Merges two subarrays of a[]
    void merge(int a[], int l, int m, int r)
    {

          int n1 = m - l + 1;
        int n2 = r - m;

        int L[] = new int[n1];
        int R[] = new int[n2];

        for (int i = 0; i < n1; ++i)
            L[i] = a[l + i];

          for (int j = 0; j < n2; ++j)
            R[j] = a[m + 1 + j];

        // Merge the temp arrays
        // Initial indexes of first and second subarrays
        int i = 0, j = 0;

        int k = l;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                a[k] = L[i];
                i++;
            }
            else {
                a[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            a[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            a[k] = R[j];
            j++;
            k++;
        }
    }

    // Main function that sorts a[l..r] using
    // merge()
    void sort(int a[], int l, int r)
    {
        if (l < r) {
          
            int m = (l + r) / 2;

            // Sort first and second halves
            sort(a, l, m);
            sort(a, m + 1, r);

            // Merge the sorted halves
            merge(a, l, m, r);
        }
    }

    // Driver method
    public static void main(String args[])
    {
        int a[] = { 12, 11, 13, 5, 6, 7 };
	int a2[] = { 100, 23, 1237, 6, 10, 1};
	int a3[] = { 0, 4, 3, 9, 1, 10, 10 };

        // Calling of Merge Sort for array 1
        merge_sort ob = new merge_sort();

	System.out.print("Test case 1: ");
	for(int i = 0; i < a.length; ++i)
                System.out.print(a[i] + " ");

	System.out.print("\nSorted: ");
	ob.sort(a, 0, a.length - 1);
        for (int i = 0; i < a.length; ++i)
            System.out.print(a[i] + " ");
	System.out.println();

	// Print array 2
	System.out.print("\nTest case 2: ");
        for(int i = 0; i < a2.length; ++i)
                System.out.print(a2[i] + " ");

        System.out.print("\nSorted: ");
        ob.sort(a2, 0, a2.length - 1);
        for (int i = 0; i < a2.length; ++i)
            System.out.print(a2[i] + " ");
        System.out.println();

	// Print array 3
	System.out.print("\nTest case 3: ");
        for(int i = 0; i < a3.length; ++i)
                System.out.print(a3[i] + " ");

        System.out.print("\nSorted: ");
        ob.sort(a3, 0, a3.length - 1);
        for (int i = 0; i < a3.length; ++i)
            System.out.print(a3[i] + " ");
        System.out.println();

    }
}
// Updated class names

File: Python/merge_sort.py
def merge(arr, l, m, r):
    n1 = m - l + 1
    n2 = r - m

    L = [0]*(n1)
    R = [0]*(n2)

    for i in range(0, n1):
        L[i] = arr[l + i]

    for j in range(0, n2):
        R[j] = arr[m + 1 + j]

    i = j = 0
    k = l

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i+=1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

def merge_sort(arr, l, r):
    if l < r:
        m = l+(r-l)//2

        merge_sort(arr, l, m)
        merge_sort(arr, m+1, r)
        merge(arr, l, m, r)

if __name__ == "__main__":
    arr = [12, 11, 13, 5, 6, 7]
    n = len(arr)
    print()
    print(f"\nStarting array : {arr}")

    merge_sort(arr, 0, n-1)

    print(f"Sorted array   : {arr}\n")